datasource/ethereum/src/block_ingestor.rs

```diff
@@ -0,0 +1,179 @@
+    ancestor_count: u64,
+    logger: slog::Logger,
+    polling_interval: Duration,
+}
+
+impl<S, T> BlockIngestor<S, T>
+where
+    S: BlockStore + Send + 'static,
+    T: BatchTransport + Send + Sync + Debug + Clone + 'static,
+    <T as Transport>::Out: Send,
+    <T as BatchTransport>::Batch: Send,
+{
+    pub fn new(
+        store: Arc<Mutex<S>>,
+        network_name: String,
+        web3_transport: T,
+        ancestor_count: u64,
+        logger: slog::Logger,
+        polling_interval: Duration,
+    ) -> Result<BlockIngestor<S, T>, Error> {
+        // Add a head block pointer for this network name if one does not already exist
+        store.lock().unwrap().add_network_if_missing(&network_name)?;
+
+        Ok(BlockIngestor {
+            store,
+            network_name,
+            web3_transport,
+            ancestor_count,
+            logger: logger.new(o!("component" => "BlockIngestor")),
+            polling_interval,
+        })
+    }
+
+    pub fn into_polling_stream(self) -> impl Future<Item = (), Error = ()> {
+        // Currently, there is no way to stop block ingestion, so just leak self
+        let static_self: &'static _ = Box::leak(Box::new(self));
+
+        // Create stream that emits at polling interval
+        tokio::timer::Interval::new(Instant::now(), static_self.polling_interval)
+            .map_err(move |e| {
+                error!(static_self.logger, "timer::Interval failed: {:?}", e);
+            })
+            .for_each(move |_| {
+                // Attempt to poll
+                static_self.do_poll().then(move |result| {
+                    if let Err(e) = result {
+                        // Some polls will fail due to transient issues
+                        warn!(
+                            static_self.logger,
+                            "failed to poll for latest block: {:?}", e
+                        );
+                    }
+
+                    // Continue polling even if polling failed
+                    future::ok(())
+                })
+            })
+    }
+
+    fn do_poll<'a>(&'a self) -> impl Future<Item = (), Error = Error> + 'a {
+        // Ask for latest block from Ethereum node
+        self.get_latest_block()
+            .and_then(move |latest_block: Block<Transaction>| {
+                // Store latest block in block store.
+                // Might be a no-op if latest block is one that we have seen.
+                // ingest_blocks will return a (potentially incomplete) list of blocks that are
+                // missing.
+                self.ingest_blocks(stream::once(Ok(latest_block)))
+            })
+            .and_then(move |missing_block_hashes| {
+                // Repeatedly fetch missing blocks, and ingest them.
+                // ingest_blocks will continue to tell us about more missing blocks until we have
+                // filled in all missing pieces of the blockchain (that we care about).
+                future::loop_fn(
```

Comment:
I still don't understand how this can loop more than twice, but that's something I can ask Tim later if you are also not sure of what's going on.

Reply:
[This comment](https://github.com/graphprotocol/graph-node/pull/309#discussion_r214336772) is the understanding I had from reading the code, if it's correct this should probably be refactored for clarity. I'll catch @timmclean later to discuss this so that more than one person has good grip of what's going on here.

Reply:
Ok, please do. In the meantime, we'll merge the PR to get this in.

Reply:
Yup, Jannis' description is correct. Leo's totally right, this needs to be clearer. I filed #336 to track documenting this. Leo: let's connect when I'm back next week and find a way to make this clearer :)